docs/dev/opinionated-classpath.md

```diff
@@ -0,0 +1,42 @@
+# Opinionated Java Classpath Container
+
+This document describes a proposal for a new way of calculating the Java classpath in Eclipse for Bazel projects.
+
+## Problem Statement
+
+Bazel dependencies for Java targets are listed in `BUILD.bazel` files (build files).
+The classpath in Eclipse is computed from these entries in build files.
+
+The dependencies need to be specified very precise, i.e. Bazel does not allow transitive dependencies.
+
+When working with third-party libraries (eg., Google Guava) it's a trivial task to modify the build files and add necessary entries.
+
+In order for any Bazel project to excel at incremental compilation, fine grained packages are recommended, i.e. Bazel packages should be defined for individual Java packages.
+When working with such fine grained Bazel packages within a large project it become a painful and diligent exercise.
+
+Multiple questions need to be answered before the build file can be updated:
+- *Where is the dependency coming from?*
+- *What is the Bazel target name of a third-party library?*
+- *Which libraries are available for the package?*
+- *Is the target visible to the package?*
+
+
+## Proposed Solution
+
+The opinionated classpath container addresses the problem by reversing it.
+Instead of reading a Bazel Java package's dependencies from the build file the build file will be populated from what the Bazel package actually uses.
```

Comment:
Simple textual parsing can give us the imports, but we will need to do deeper analysis to pick up class references in the body (e.g. ```java.util.Map x = new java.util.HashMap();``` )